scripts/conversion_script.py

```diff
@@ -289,6 +302,16 @@ def cleanup(required_files):
             os.remove(required_file.path)
         _create_or_restore_backup_file(required_file)
 
+    if undo_last_yum_transaction:
+        output, returncode = run_subprocess(
+            ["/usr/bin/yum", "history", "undo", "last"],
+        )
+        if returncode:
+            print(
+                "Undo of last yum transaction failed with exist status '%s' and output '%s'"
+                % (returncode, output)
+            )
+
```

Comment:
Oh, that might be tricky... Have you checked if the last transaction might not be something executed by convert2rhel itself?

Reply:
I would rather execute a `yum remove convert2rhel` instead.

Reply:
The reason is that @pholica mentioned we should not only remove the installed convert2rhel but also all the deps that came with it. A modern yum removes unused deps automatically but yum does not do that.

I've now figured that we can't use the `yum history undo last`. I've tried manually after the `convert2rhel analyze` finished and it started removing subscription-manager. That's because convert2rhel installed it. We clearly can't rely on the fact that the last yum transaction was the installation of convert2rhel.

I guess what you could do is to get the sequence number of the yum transaction in yum history right after installing convert2rhel and then run `# yum history undo <number>`.

Reply:
I'm sorry, but there must be some confusion. This PR is about changing the `convert2rhel-worker-scripts` and not convert2rhel. The flow is following:

1. User installs subscription-manager and other packages
2. User registers the systems using rhc
3. User starts a task (covered by this code)
    1. Task installs conver2rhel
    2. Task runs convert2rhel
    3. Task runs `yum history undo last`
Within this flow, the convert2rhel doesn't perform any yum transaction. It is true however that the `last` was merely a shortcut which should not be blindly used. The last transaction id (after installing the convert2rhel can be obtained by something like: `yum history info | awk -F'[[:space:]]*:[[:space:]]' '$1 == "Transaction ID" {print $2}'`).

I'd like to warn about the `yum history list convert2rhel` approach as that could be a very problematic if convert2rhel was previously installed by the user (possibly with other packages as part of the transaction) and undoing this user transaction could lead to another undesired situation.

Reply:
Resolved other conversation because I just pushed new changes, so the logic is following:

- cleanup after pre-conversion
  - for both fails and successes - if convert2rhel existed on system before it is kept on system (update of the pkg is not rolled back), else cleaned up
-  cleanup after conversion
  - for fails - if convert2rhel existed on system before it is kept on system (update of the pkg is not rolled back), else cleaned up
  - for success - pkg is always kept because of statistics with the note that this can be removed after we do not rely on that fact

**Now in all cases the repo is removed - if one existed before script run it will be restored** - tell me if we want to keep it in case of the successful conversion (its minor change)

How does the cleanup work with the newest changes:

After pkg installation is done I call function that takes output of `yum history list $PKG`, and last transaction ID is taken (last line) and that id is then used in cleanup function to roll back the transaction (it is extensible for future if we needed to roll back more transactions, even if we decided to roll back the updates).
And I know that the last transaction (the one we undo) should be ours because we do the pkg installed check first, then installation/update based on that and then immediately after save the last transaction for that pkg for end cleanup. So it shouldn't happen that we undo something that isn't ours.

Only edge case is that *IF* we ever decided to roll back even the updates of c2r pkg we would have to be sure that the package was updated in the first place and add transaction_id only if it really was. But we don't donwgrade the pkg right now, if it was on system we try to update it and then keep it on system.

Reply:
@pholica Please take a look, no code review required, but check the logic in previous comment, @r0x0d  and @bocekm  please check the code as well, I added some unit tests which should make sure it work's as I intended.

Reply:
Hmm, I'm not sure about the logic. It sounds like it could do the trick (if I ignore the decision how to handle the inconsistency after successful conversion).